core/shared/src/main/scala/io/circe/Error.scala

```diff
@@ -27,6 +34,44 @@ final object DecodingFailure {
     case (DecodingFailure(m1, h1), DecodingFailure(m2, h2)) =>
       m1 == m2 && Eq[List[HistoryOp]].eqv(h1, h2)
   }
+
+  /**
+    * Creates compact, human readable string representations for DecodingFailure
+    * Cursor history is represented as JS style selections, i.e. ".foo.bar[3]"
+    */
+  implicit final val showDecodingFailure: Show[DecodingFailure] = Show.show { failure =>
+
+    /** Represents JS style selections into JSON structure */
+    sealed trait Selection
```

Comment:
This is a relatively minor issue, but I think I'd like to try to avoid defining types in methods whenever possible—these get stuck in `DecodingFailure` and leak into its public API. We could either bump them up into `DecodingFailure` and make them `private[this]` there, or we could define `Show` as an anonymous class instance (I don't have a strong opinion either way).